File: 2025/03 March/31 Maximize partitions in a String/main.py
# Link: https://www.geeksforgeeks.org/problems/maximize-partitions-in-a-string/1


class Solution:
    def maxPartitions(self, s: str) -> int:
        cnt = 0
        n = len(s)
        mp = {char: i for i, char in enumerate(s)}

        ans = -1
        for i in range(n):
            ans = max(ans, mp[s[i]])

            if ans == i:
                cnt += 1
        return cnt


if __name__ == "__main__":
    sol = Solution()
    s = "acbbcc"
    print(sol.maxPartitions(s))

    s = "ababcbacadefegdehijhklij"
    print(sol.maxPartitions(s))

    s = "aaa"
    print(sol.maxPartitions(s))

File: 2025/03 March/31 Maximize partitions in a String/main.cpp
// Link: https://www.geeksforgeeks.org/problems/maximize-partitions-in-a-string/1

#include <iostream>
#include <unordered_map>
#include <cmath>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int maxPartitions(string &s)
    {
        int cnt = 0;
        int n = s.size();
        unordered_map<char, int> mp;
        for (int i = 0; i < n; i++)
            mp[s[i]] = i;

        int ans = -1;
        for (int i = 0; i < n; i++)
        {
            ans = fmax(ans, mp[s[i]]);

            if (ans == i)
                cnt++;
        }
        return cnt;
    }
};

int main()
{
    Solution sol;
    string s = "acbbcc";
    cout << sol.maxPartitions(s) << endl;

    s = "ababcbacadefegdehijhklij";
    cout << sol.maxPartitions(s) << endl;

    s = "aaa";
    cout << sol.maxPartitions(s) << endl;

    return 0;
}